include/device.h

```diff
@@ -254,6 +271,7 @@ struct device {
 	const void *config_info;
 	const void *driver_api;
 	void * const driver_data;
+	struct device_context * const driver_context;
```

Comment:
I'd just call it context (imo all these attributes should be renamed: config_info -> config, driver_api -> api, driver_data -> data), but that's a side note.

What's the benefit exposing it that way? (adding a pointer in the struct, instead of a fully separate array as in #24873).  It's a trade-off between ROM occupation and runtime computation, anything else than that?

In the very end, we should merge driver_data and driver_context, and provide the means for the drivers to access the data seamlessly (via a macro or else). I just did not go myself in that path in my RFC because it generates a lot of changes in drivers

Reply:
Exactly as you say: the right way is to put the common context into a prefix of the driver-specific data.  That can't be done if they're separate objects.

This is a prototype using a different object relationship than yours; we don't know yet what the best path will be.  Better to try several and compare their pros and cons.

Reply:
>  That can't be done if they're separate objects.

Of course it can be. Current RFCs (mine and yours) are only trying to solve the device instance as constant. 

Once done _then_ we can move on with modifying driver_data to provide also a driver context. 

But, as I said in my other comment, if the goal is _only_ to solve the device instances to be constants, there is 0 point to introduce right now a struct device_context since an atomic bits array would just do the work.
In my RFC I did introduce such context because of all the other features (sync/lock and status reporting), not because of const-ifying devices. 

So maybe we should really proceed step by steps: 
- let's fix the device constant by having such atomic bits (and let device_get_binding() check these when relevant)
- let's add a struct device_context for everything else that would come afterwards (sync/lock & status reporting).

As you already mentioned in relevant PRs, you are not convinced about adding such sync/lock and status report features... so adding a device context right now would be a bad idea then.

Reply:
You're right, I was combining multiple features.  I'll switch it to a bool `init_ok` since that's the only state we need right now.